apps/andi/lib/andi_web/live/submit_live_view/upload_data_dictionary.ex

```diff
@@ -0,0 +1,248 @@
+                <div class="loader data-dictionary-form__loader data-dictionary-form__loader--<%= loader_visibility %>"></div>
+              </div>
+          </div>
+
+            <div class="edit-button-group form-grid">
+              <div class="edit-button-group__cancel-btn">
+                <a href="#data-dictionary-form" id="back-button" class="btn btn--back btn--large" phx-click="toggle-component-visibility" phx-value-component-expand="metadata_form">Back</a>
+                <button type="button" class="btn btn--large" phx-click="cancel-edit">Cancel</button>
+              </div>
+
+              <div class="edit-button-group__save-btn">
+                <a href="#review_submission" id="next-button" class="btn btn--next btn--large btn--action" phx-click="toggle-component-visibility" phx-value-component-expand="data_dictionary_form">Next</a>
+              </div>
+            </div>
+          </div>
+        </form>
+      </div>
+    </div>
+    """
+  end
+
+  def handle_event("validate", _, socket) do
+    {:noreply, socket}
+  end
+
+  def handle_info(
+        %{topic: "toggle-visibility", payload: %{expand: "metadata_form", dataset_id: dataset_id}},
+        %{assigns: %{dataset_id: dataset_id}} = socket
+      ) do
+    {:noreply, assign(socket, visibility: "expanded") |> update_validation_status()}
+  end
+
+  def handle_info(%{topic: "toggle-visibility"}, socket) do
+    {:noreply, socket}
+  end
+
+  def handle_event("file_upload", %{"fileType" => file_type}, socket)
+      when file_type not in ["text/csv", "application/json", "application/vnd.ms-excel"] do
+    new_changeset =
+      socket.assigns.changeset
+      |> reset_changeset_errors()
+      |> Ecto.Changeset.add_error(:datasetLink, "File type must be CSV or JSON")
+      |> Map.put(:action, :update)
+
+    {:noreply, assign(socket, changeset: new_changeset, loading_schema: false)}
+  end
+
+  def handle_event("file_upload", %{"fileSize" => file_size}, socket) when file_size > 200_000_000 do
+    new_changeset =
+      socket.assigns.changeset
+      |> reset_changeset_errors()
+      |> Ecto.Changeset.add_error(:datasetLink, "File size must be less than 200MB")
+      |> Map.put(:action, :update)
+
+    {:noreply, assign(socket, changeset: new_changeset, loading_schema: false)}
+  end
+
+  def handle_event("file_upload", %{"file" => file, "fileName" => file_name, "fileType" => file_type}, socket)
+      when file_type in ["text/csv", "application/vnd.ms-excel"] do
+    dataset_id = socket.assigns.dataset_id
+    dataset_link = "#{@bucket_path}#{dataset_id}/#{file_name}"
+
+    with {:ok, presigned_url} <- presigned_url(dataset_id, file_name),
+         {:ok, _} <- upload_sample_dataset(presigned_url, file, "text/csv") do
+      %{datasetLink: dataset_link}
+      |> DatasetLinkFormSchema.changeset_from_form_data()
+      |> send_dataset_link_status(socket)
+      |> complete_validation(socket)
```

Comment:
this might be better broken out into its own function, like the `complete_validation` functions in some of our other liveview pages